### PR TITLE
PKGBUILD file for ArchUserRepo submission

### PR DESCRIPTION
**Pull request by CosminGGeorgescu**
As someone suggested in [this issue](https://github.com/sagiegurari/duckscript/issues/274), I've added the PKGBUILD file and necessary for submitting to the Arch User Repository. 

&nbsp;&nbsp;&nbsp;&nbsp;Alternatively, if you will merge PR, users will be able to just `makepkg -ci` inside the cloned directory and build the package locally followed by a `pacman -U` installation, though, since `cargo` is a make dependency and they will have already cloned the repo, it is of questionable use as they could just take the `cargo install` route.

&nbsp;&nbsp;&nbsp;&nbsp;Just wanted to check in with you regarding the project building and everything before submitting the file to AUR. As per [check() doc](https://wiki.archlinux.org/title/creating_packages#check()), this function could be removed. Since this a release package, this function currently serves to locally check that you did not release a broken version. I would remove it as it decreases installation time with little to no dangers, let me know what you think.

### CODE REVIEW

**Comment by sagiegurari** [comment]
Thanks, actually cargo is not a dependency. if it's not there, some functionality won't work like automatic install of cargo plugins.

**Comment by sagiegurari** [comment]
ok, i see what i'm missing. cargo is used by the pkgbuild.

File: PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	cd "$pkgname-$pkgver"
+	export RUSTUP_TOOLCHAIN=stable
+	export CARGO_TARGET_DIR=target
+	cargo build --frozen --workspace --release --all-features
+}
+
+check(){
+	cd "$pkgname-$pkgver"
+	export RUSTUP_TOOLCHAIN=stable
+	cargo test --frozen --workspace --all-features
```
Comment (sagiegurari) [inline PKGBUILD:38]: some tests are not thread safe and some are.
duckscript by default uses cargo-make which splits the run to single thread and multi thread.
single thread tests are by default ignored (which is why your test command works), so what cargo-make does is 
1. run cargo test on everything so its multi thread and single thread tests are ignored
2. sets RUST_TEST_THREADS=1 and runs cargo test with --ignored

so if you want to run ALL tests, thats what you got to do.
if you want to run with partial coverage, you can keep what you did.

wdyt?

Comment (CosminGGeorgescu) [reply PKGBUILD:38]: Sorry about the delay; had some finals.
I would go with running the full test suite.

**Comment by sagiegurari** [comment]
@CosminGGeorgescu sorry for late reply.
Many things in these files are hardcoded (hashes, version numbers) and would require manual work to setup.
won't it be better to setup some script to update a template and generate these?

this library uses cargo-make so it could be a task there (gen-pckbuild) that i can latter easily hook into my release flow

**Comment by CosminGGeorgescu** [comment]
> Many things in these files are hardcoded (hashes, version numbers) and would require manual work to setup.
won't it be better to setup some script to update a template and generate these?

That's how they should be. The maintainer (me) will take care of updating them every time you publish a new release. Just give me a few days to take care of an exam and I will finish everything.

File: PKGBUILD (hunk 1 above)
Comment (CosminGGeorgescu) [reply PKGBUILD:38]: I've pushed the commit that should address this. So now,
`cargo test --frozen --workspace --all-features | grep ignored` will report 
>test result: ok. 179 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
test lib_test::test_scripts ... ignored
test sdk::std::env::set_current_directory::mod_test::run_no_arguments ... ignored
test sdk::std::env::set_current_directory::mod_test::run_specific_directory ... ignored
test sdk::std::env::set_current_directory::mod_test::run_to_directory_does_not_exist ... ignored
test result: ok. 823 passed; 0 failed; 4 ignored; 0 measured; 0 filtered out; finished in 0.92s
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s 

and `cargo test --frozen --workspace --all-features -- --ignored --test-threads=1 | grep -w test` will report the remaining tests
>test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 179 filtered out; finished in 0.00s
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
test lib_test::test_scripts ... hello world
test sdk::std::env::set_current_directory::mod_test::run_no_arguments ... ok
test sdk::std::env::set_current_directory::mod_test::run_specific_directory ... ok
test sdk::std::env::set_current_directory::mod_test::run_to_directory_does_not_exist ... ok
test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 823 filtered out; finished in 6.77s
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

In total, 1006 tests were run.
Let me know if I missed anything.

**Comment by sagiegurari** [comment]
@CosminGGeorgescu this is still a bit strange to me. I've got no experience with pkgbuild files, but seems like this could be a template and a script to automate the updating of the template (via cargo-make preferred).
relying on you to update on each new release (thanks) doesn't scale and best to have instructions (script) to do that. no?
also do you submit this anywhere?

**Comment by CosminGGeorgescu** [comment]
> also do you submit this anywhere?

https://aur.archlinux.org/packages/duckscript .
Here it is, sorry for the delay.
Now, if there will be enough people that find this useful and vote for it, it will reach the official Archlinux package database and people will be able to install it using `pacman`.

> but seems like this could be a template and a script to automate the updating of the template

Yes, it is and it's called [updpkgsums](https://man.archlinux.org/man/updpkgsums.8.en) as the checksums are the only variables that should be updated automatically. 

> doesn't scale and best to have instructions (script) to do that. no?

I'm completely with you on this one, but it's just a case of idiomatic thing. Normally maintainers have a small script to update the pkgfiles and, as this is my first package maintain, I see no reason to unnecessarily add to your `cargo-make` something that would take me 3 mins tops by hand.

File: PKGBUILD [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+source=("$pkgname-$pkgver.tar.gz::https://github.com/sagiegurari/duckscript/archive/$pkgver.tar.gz")
+
+sha512sums=("7ffe4ad2d1bb54753d7c262981bf31394404dd71a75919188a5f658de39b9b9215d057da382e220433d866369417e74feaef3ab4ac80df5214b046284f5f840d")
+
+b2sums=("869859f0d7e21abf1f3e1041e6f7056c25d4e6341bab19788c95c609da60cff5add66ec113aa1f2ff9557596537728a6255ec65c280668d9d3e8e785d042f616")
+
+prepare(){
+	export RUSTUP_TOOLCHAIN=stable
+	cargo fetch --locked --target "$arch-unknown-linux-gnu"
```
Comment (CosminGGeorgescu) [inline PKGBUILD:25]: It crossed my mind just now, should the libc be `musl` or `gnu` ? For each release, I've noticed you only publish the `x86_64-unknown-linux-musl` version and never a `gnu` one.

Comment (sagiegurari) [reply PKGBUILD:25]: i don't mind publishing all of them.

Comment (CosminGGeorgescu) [reply PKGBUILD:25]: Ah, only if you want to; it doesn't help the script in any way. Thought there might be a performance reason or something you found.

**Comment by CosminGGeorgescu** [comment]
Think I'll close this PR in a few days as there is nothing more to do. If you've anything more to add, let me know. I'll send a new PR next week for the install. Hopefully, now more people will able to reach your project and if it'll be popular enough with Arch folk, it will reach the official package database. Thank you for everything !

**Comment by sagiegurari** [comment]
the only thing 'im missing (we talked about it) is a script or something that creates the file with the checksums.
or at least, put some instructions at the top of the PKGBUILD file in a comment

**Pull request closed without merging**